cdc reports a checkpoint regress error

### What did you do?

cdc should work fine.

### What did you expect to see?

_No response_

### What did you see instead?

[2023/06/13 23:49:42.726 +08:00] [ERROR] [processor.go:525] ["run processor failed"] [capture=21928511-99c7-421a-92a4-70d4925080a4] [namespace=default] [changefeed=x] [error="[CDC:ErrInvalidCheckpointTs]checkpointTs(442150782138318853) should not larger than resolvedTs(442150782124949509)"] [errorVerbose="[CDC:ErrInvalidCheckpointTs]checkpointTs(442150782138318853) should not larger than resolvedTs(442150782124949509)\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/normalize.go:164\ngithub.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent.(*agent).handleMessageHeartbeat\n\tgithub.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent/agent.go:295\ngithub.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent.(*agent).handleMessage\n\tgithub.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent/agent.go:257\ngithub.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent.(*agent).Tick\n\tgithub.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent/agent.go:209\ngithub.com/pingcap/tiflow/cdc/processor.(*processor).tick\n\tgithub.com/pingcap/tiflow/cdc/processor/processor.go:583\ngithub.com/pingcap/tiflow/cdc/processor.(*processor).Tick\n\tgithub.com/pingcap/tiflow/cdc/processor/processor.go:471\ngithub.com/pingcap/tiflow/cdc/processor.(*managerImpl).Tick\n\tgithub.com/pingcap/tiflow/cdc/processor/manager.go:141\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).Run\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:290\ngithub.com/pingcap/tiflow/cdc/capture.(*captureImpl).runEtcdWorker\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:523\ngithub.com/pingcap/tiflow/cdc/capture.(*captureImpl).run.func4\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:354\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.2.0/errgroup/errgroup.go:75\nruntime.goexit\n\truntime/asm_amd64.s:1598"]

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
(paste TiDB cluster version here)
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):

```console
(paste TiCDC version here)
```

--- a/cdc/processor/processor.go
+++ b/cdc/processor/processor.go
@@ -373,8 +373,8 @@ func (p *processor) getStatsFromSourceManagerAndSinkManager(
 		ResolvedTs:   sortStats.ReceivedMaxResolvedTs,
 	}
 	stats.StageCheckpoints["sorter-egress"] = tablepb.Checkpoint{
-		CheckpointTs: sinkStats.ReceivedMaxCommitTs,
-		ResolvedTs:   sinkStats.ReceivedMaxCommitTs,
+		CheckpointTs: sinkStats.ResolvedTs,
+		ResolvedTs:   sinkStats.ResolvedTs,
 	}
 
 	return stats

--- a/cdc/processor/processor_test.go
+++ b/cdc/processor/processor_test.go
@@ -220,9 +220,8 @@ func TestTableExecutorAddingTableIndirectly(t *testing.T) {
 	p.sinkManager.r.UpdateBarrierTs(20, nil)
 	stats := p.sinkManager.r.GetTableStats(span)
 	require.Equal(t, model.Ts(20), stats.CheckpointTs)
+	require.Equal(t, model.Ts(20), stats.ResolvedTs)
 	require.Equal(t, model.Ts(20), stats.BarrierTs)
-	require.Equal(t, model.Ts(0), stats.ReceivedMaxCommitTs)
-	require.Equal(t, model.Ts(20), stats.ReceivedMaxResolvedTs)
 	require.Len(t, p.sinkManager.r.GetAllCurrentTableSpans(), 1)
 	require.Equal(t, 1, p.sinkManager.r.GetAllCurrentTableSpansCount())
 
@@ -259,9 +258,8 @@ func TestTableExecutorAddingTableIndirectly(t *testing.T) {
 	require.True(t, ok)
 	stats = p.sinkManager.r.GetTableStats(span)
 	require.Equal(t, model.Ts(20), stats.CheckpointTs)
+	require.Equal(t, model.Ts(101), stats.ResolvedTs)
 	require.Equal(t, model.Ts(20), stats.BarrierTs)
-	require.Equal(t, model.Ts(0), stats.ReceivedMaxCommitTs)
-	require.Equal(t, model.Ts(101), stats.ReceivedMaxResolvedTs)
 
 	// Start to replicate table-1.
 	ok, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(1), 30, false)

--- a/cdc/processor/sinkmanager/manager.go
+++ b/cdc/processor/sinkmanager/manager.go
@@ -56,9 +56,6 @@ type TableStats struct {
 	CheckpointTs model.Ts
 	ResolvedTs   model.Ts
 	BarrierTs    model.Ts
-	// From sorter.
-	ReceivedMaxCommitTs   model.Ts
-	ReceivedMaxResolvedTs model.Ts
 }
 
 // SinkManager is the implementation of SinkManager.
@@ -934,11 +931,11 @@ func (m *SinkManager) GetTableStats(span tablepb.Span) TableStats {
 	if m.redoDMLMgr != nil {
 		resolvedTs = m.redoDMLMgr.GetResolvedTs(span)
 	} else {
-		resolvedTs = m.sourceManager.GetTableResolvedTs(span)
+		resolvedTs = tableSink.getReceivedSorterResolvedTs()
 	}
 
 	if resolvedTs < checkpointTs.ResolvedMark() {
-		log.Error("sinkManager: resolved ts should not less than checkpoint ts",
+		log.Panic("sinkManager: resolved ts should not less than checkpoint ts",
 			zap.String("namespace", m.changefeedID.Namespace),
 			zap.String("changefeed", m.changefeedID.ID),
 			zap.Stringer("span", &span),
@@ -946,11 +943,9 @@ func (m *SinkManager) GetTableStats(span tablepb.Span) TableStats {
 			zap.Any("checkpointTs", checkpointTs))
 	}
 	return TableStats{
-		CheckpointTs:          checkpointTs.ResolvedMark(),
-		ResolvedTs:            resolvedTs,
-		BarrierTs:             tableSink.barrierTs.Load(),
-		ReceivedMaxCommitTs:   tableSink.getReceivedSorterCommitTs(),
-		ReceivedMaxResolvedTs: tableSink.getReceivedSorterResolvedTs(),
+		CheckpointTs: checkpointTs.ResolvedMark(),
+		ResolvedTs:   resolvedTs,
+		BarrierTs:    tableSink.barrierTs.Load(),
 	}
 }
 

--- a/cdc/processor/sinkmanager/manager.go
+++ b/cdc/processor/sinkmanager/manager.go
@@ -935,7 +935,7 @@ func (m *SinkManager) GetTableStats(span tablepb.Span) TableStats {
 	}
 
 	if resolvedTs < checkpointTs.ResolvedMark() {
-		log.Panic("sinkManager: resolved ts should not less than checkpoint ts",
+		log.Error("sinkManager: resolved ts should not less than checkpoint ts",
 			zap.String("namespace", m.changefeedID.Namespace),
 			zap.String("changefeed", m.changefeedID.ID),
 			zap.Stringer("span", &span),

--- a/cdc/processor/sinkmanager/redo_log_worker.go
+++ b/cdc/processor/sinkmanager/redo_log_worker.go
@@ -89,7 +89,6 @@ func (w *redoWorker) handleTask(ctx context.Context, task *redoTask) (finalErr e
 	cachedSize := uint64(0)
 
 	defer func() {
-		task.tableSink.updateReceivedSorterCommitTs(advancer.lastTxnCommitTs)
 		eventCount := newRangeEventCount(advancer.lastPos, allEventCount)
 		task.tableSink.updateRangeEventCounts(eventCount)
 

--- a/cdc/processor/sinkmanager/table_sink_worker.go
+++ b/cdc/processor/sinkmanager/table_sink_worker.go
@@ -155,7 +155,6 @@ func (w *sinkWorker) handleTask(ctx context.Context, task *sinkTask) (finalErr e
 
 		// If eventCache is nil, update sorter commit ts and range event count.
 		if w.eventCache == nil {
-			task.tableSink.updateReceivedSorterCommitTs(advancer.currTxnCommitTs)
 			eventCount := newRangeEventCount(advancer.lastPos, allEventCount)
 			task.tableSink.updateRangeEventCounts(eventCount)
 		}

--- a/cdc/processor/sinkmanager/table_sink_wrapper.go
+++ b/cdc/processor/sinkmanager/table_sink_wrapper.go
@@ -64,9 +64,6 @@ type tableSinkWrapper struct {
 	// receivedSorterResolvedTs is the resolved ts received from the sorter.
 	// We use this to advance the redo log.
 	receivedSorterResolvedTs atomic.Uint64
-	// receivedSorterCommitTs is the commit ts received from the sorter.
-	// We use this to statistics the latency of the table sorter.
-	receivedSorterCommitTs atomic.Uint64
 	// receivedEventCount is the number of events received from the sorter.
 	receivedEventCount atomic.Int64
 
@@ -189,12 +186,6 @@ func (t *tableSinkWrapper) updateReceivedSorterResolvedTs(ts model.Ts) {
 	}
 }
 
-func (t *tableSinkWrapper) updateReceivedSorterCommitTs(ts model.Ts) {
-	if ts > t.receivedSorterCommitTs.Load() {
-		t.receivedSorterCommitTs.Store(ts)
-	}
-}
-
 func (t *tableSinkWrapper) updateResolvedTs(ts model.ResolvedTs) error {
 	t.tableSinkMu.RLock()
 	defer t.tableSinkMu.RUnlock()
@@ -223,10 +214,6 @@ func (t *tableSinkWrapper) getCheckpointTs() model.ResolvedTs {
 
 func (t *tableSinkWrapper) getReceivedSorterResolvedTs() model.Ts {
 	return t.receivedSorterResolvedTs.Load()
-}
-
-func (t *tableSinkWrapper) getReceivedSorterCommitTs() model.Ts {
-	return t.receivedSorterCommitTs.Load()
 }
 
 func (t *tableSinkWrapper) getReceivedEventCount() int64 {

--- a/cdc/processor/sourcemanager/engine/engine.go
+++ b/cdc/processor/sourcemanager/engine/engine.go
@@ -38,9 +38,6 @@ type SortEngine interface {
 	// events are available for fetching, OnResolve is what you want.
 	Add(span tablepb.Span, events ...*model.PolymorphicEvent)
 
-	// GetResolvedTs gets resolved timestamp of the given table.
-	GetResolvedTs(span tablepb.Span) model.Ts
-
 	// OnResolve pushes action into SortEngine's hook list, which
 	// will be called after any events are resolved.
 	OnResolve(action func(tablepb.Span, model.Ts))

--- a/cdc/processor/sourcemanager/engine/memory/event_sorter.go
+++ b/cdc/processor/sourcemanager/engine/memory/event_sorter.go
@@ -214,12 +214,6 @@ func (s *tableSorter) add(events ...*model.PolymorphicEvent) (resolvedTs model.T
 	return
 }
 
-func (s *tableSorter) getResolvedTs() model.Ts {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
-	return *s.resolvedTs
-}
-
 func (s *tableSorter) fetch(
 	span tablepb.Span, lowerBound, upperBound engine.Position,
 ) engine.EventIterator {

--- a/cdc/processor/sourcemanager/engine/memory/event_sorter.go
+++ b/cdc/processor/sourcemanager/engine/memory/event_sorter.go
@@ -89,16 +89,6 @@ func (s *EventSorter) Add(span tablepb.Span, events ...*model.PolymorphicEvent) 
 	}
 }
 
-// GetResolvedTs implements engine.SortEngine.
-func (s *EventSorter) GetResolvedTs(span tablepb.Span) model.Ts {
-	value, exists := s.tables.Load(span)
-	if !exists {
-		log.Panic("get resolved ts from an unexist table", zap.Stringer("span", &span))
-	}
-
-	return value.(*tableSorter).getResolvedTs()
-}
-
 // OnResolve implements engine.SortEngine.
 func (s *EventSorter) OnResolve(action func(tablepb.Span, model.Ts)) {
 	s.mu.Lock()

--- a/cdc/processor/sourcemanager/engine/mock/engine_mock.go
+++ b/cdc/processor/sourcemanager/engine/mock/engine_mock.go
@@ -135,20 +135,6 @@ func (mr *MockSortEngineMockRecorder) FetchByTable(span, lowerBound, upperBound 
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "FetchByTable", reflect.TypeOf((*MockSortEngine)(nil).FetchByTable), span, lowerBound, upperBound)
 }
 
-// GetResolvedTs mocks base method.
-func (m *MockSortEngine) GetResolvedTs(span tablepb.Span) model.Ts {
-	m.ctrl.T.Helper()
-	ret := m.ctrl.Call(m, "GetResolvedTs", span)
-	ret0, _ := ret[0].(model.Ts)
-	return ret0
-}
-
-// GetResolvedTs indicates an expected call of GetResolvedTs.
-func (mr *MockSortEngineMockRecorder) GetResolvedTs(span interface{}) *gomock.Call {
-	mr.mock.ctrl.T.Helper()
-	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "GetResolvedTs", reflect.TypeOf((*MockSortEngine)(nil).GetResolvedTs), span)
-}
-
 // GetStatsByTable mocks base method.
 func (m *MockSortEngine) GetStatsByTable(span tablepb.Span) engine.TableStats {
 	m.ctrl.T.Helper()

--- a/cdc/processor/sourcemanager/engine/pebble/event_sorter.go
+++ b/cdc/processor/sourcemanager/engine/pebble/event_sorter.go
@@ -160,44 +160,23 @@ func (s *EventSorter) Add(span tablepb.Span, events ...*model.PolymorphicEvent) 
 			zap.Stringer("span", &span))
 	}
 
-	maxCommitTs := model.Ts(0)
-	maxResolvedTs := model.Ts(0)
+	maxCommitTs := state.maxReceivedCommitTs.Load()
+	maxResolvedTs := state.maxReceivedResolvedTs.Load()
 	for _, event := range events {
-		state.ch.In() <- eventWithTableID{uniqueID: state.uniqueID, span: span, event: event}
 		if event.IsResolved() {
 			if event.CRTs > maxResolvedTs {
 				maxResolvedTs = event.CRTs
+				state.maxReceivedResolvedTs.Store(maxResolvedTs)
 			}
 		} else {
 			state.receivedEvents.Add(1)
 			if event.CRTs > maxCommitTs {
 				maxCommitTs = event.CRTs
+				state.maxReceivedCommitTs.Store(maxCommitTs)
 			}
 		}
+		state.ch.In() <- eventWithTableID{uniqueID: state.uniqueID, span: span, event: event}
 	}
-
-	if maxCommitTs > state.maxReceivedCommitTs.Load() {
-		state.maxReceivedCommitTs.Store(maxCommitTs)
-	}
-	if maxResolvedTs > state.maxReceivedResolvedTs.Load() {
-		state.maxReceivedResolvedTs.Store(maxResolvedTs)
-	}
-}
-
-// GetResolvedTs implements engine.SortEngine.
-func (s *EventSorter) GetResolvedTs(span tablepb.Span) model.Ts {
-	s.mu.RLock()
-	state, exists := s.tables.Get(span)
-	s.mu.RUnlock()
-
-	if !exists {
-		log.Panic("get resolved ts from an non-existent table",
-			zap.String("namespace", s.changefeedID.Namespace),
-			zap.String("changefeed", s.changefeedID.ID),
-			zap.Stringer("span", &span))
-	}
-
-	return state.maxReceivedResolvedTs.Load()
 }
 
 // OnResolve implements engine.SortEngine.

--- a/cdc/processor/sourcemanager/engine/pebble/event_sorter_test.go
+++ b/cdc/processor/sourcemanager/engine/pebble/event_sorter_test.go
@@ -44,7 +44,7 @@ func TestTableOperations(t *testing.T) {
 	s.AddTable(span, 2)
 	s.AddTable(span, 1)
 
-	require.Equal(t, model.Ts(2), s.GetResolvedTs(span))
+	require.Equal(t, model.Ts(2), s.GetStatsByTable(span).ReceivedMaxResolvedTs)
 
 	s.RemoveTable(span)
 	s.RemoveTable(span)
@@ -128,7 +128,7 @@ func TestEventFetch(t *testing.T) {
 
 	s.Add(span, inputEvents...)
 	s.Add(span, model.NewResolvedPolymorphicEvent(0, 4))
-	require.Equal(t, model.Ts(4), s.GetResolvedTs(span))
+	require.Equal(t, model.Ts(4), s.GetStatsByTable(span).ReceivedMaxResolvedTs)
 
 	sortedEvents := make([]*model.PolymorphicEvent, 0, len(inputEvents))
 	sortedPositions := make([]engine.Position, 0, len(inputEvents))

--- a/cdc/processor/sourcemanager/manager.go
+++ b/cdc/processor/sourcemanager/manager.go
@@ -140,11 +140,6 @@ func (m *SourceManager) CleanByTable(span tablepb.Span, upperBound engine.Positi
 	return m.engine.CleanByTable(span, upperBound)
 }
 
-// GetTableResolvedTs returns the resolved ts of the table.
-func (m *SourceManager) GetTableResolvedTs(span tablepb.Span) model.Ts {
-	return m.engine.GetResolvedTs(span)
-}
-
 // GetTablePullerStats returns the puller stats of the table.
 func (m *SourceManager) GetTablePullerStats(span tablepb.Span) puller.Stats {
 	p, ok := m.pullers.Load(span)

